acurl/acurl/utils.py

```diff
@@ -0,0 +1,54 @@
+from collections import defaultdict
+
+
+class CaseInsensitiveDefaultDict(defaultdict):
+    # Base on https://stackoverflow.com/a/32888599 but tweaked for python3
+    # and using defaultdict as a base
+    @classmethod
+    def _k(cls, key):
+        return key.lower() if isinstance(key, str) else key
+
+    def __init__(self, *args, **kwargs):
+        super(CaseInsensitiveDefaultDict, self).__init__(*args, **kwargs)
+        self._convert_keys()
+
+    def __getitem__(self, key):
+        return super(CaseInsensitiveDefaultDict, self).__getitem__(self.__class__._k(key))
+
+    def __setitem__(self, key, value):
+        super(CaseInsensitiveDefaultDict, self).__setitem__(self.__class__._k(key), value)
+
+    def __delitem__(self, key):
+        return super(CaseInsensitiveDefaultDict, self).__delitem__(self.__class__._k(key))
+
+    def __contains__(self, key):
+        return super(CaseInsensitiveDefaultDict, self).__contains__(
+            self.__class__._k(key)
+        )
+
+    def has_key(self, key):
+        return super(CaseInsensitiveDefaultDict, self).has_key(self.__class__._k(key))
+
+    def pop(self, key, *args, **kwargs):
+        return super(CaseInsensitiveDefaultDict, self).pop(
+            self.__class__._k(key), *args, **kwargs
+        )
+
+    def get(self, key, *args, **kwargs):
+        return super(CaseInsensitiveDefaultDict, self).get(
+            self.__class__._k(key), *args, **kwargs
+        )
+
+    def setdefault(self, key, *args, **kwargs):
+        return super(CaseInsensitiveDefaultDict, self).setdefault(
+            self.__class__._k(key), *args, **kwargs
+        )
+
+    def update(self, E={}, **F):
+        super(CaseInsensitiveDefaultDict, self).update(self.__class__(E))
+        super(CaseInsensitiveDefaultDict, self).update(self.__class__(**F))
```

Comment:
```suggestion
        super(CaseInsensitiveDefaultDict, self).update(self.__class__(E, **F))
```

I think we can get away with just a single call like this.

Reply:
And on top of this, I understand E should default to None as it is optional.